### DISCUSSION TITLE
Proton VPN (paid, with port forwarding) - port shows up as open on slsk website, but still looks closed in program - anyone got Proton VPN fully working?

### DISCUSSION THREAD

**Discussion by exostalker** [nicotine-plus/nicotine-plus#2935] [Q&A] [answered]
Hey, wondering if anyone got Nicotine+ working fully with Proton VPN (paid, with port forwarding)?

When I indicate the supposedly open port from Proton VPN in the client, even though **soulseek website for checking if port is open says it is** - I can definitely tell it isn't working that way in the program itself.
 
This is quickly revealed while testing through another computer of mine with another soulseek client as I can tell there with my static connection from ISP there (since I know for certain there that if I open them there they are open - if open ports on that other computer there, I can find the files - if closed, can't - meaning that my VPN client computer has closed ports too) ; as well as one buddy of mine in the list clearly has their ports closed as I can't browse them with VPN at all (and yet I can as soon as I turn off the VPN and use my simple static connection with open ports). 
Another obvious giveaway being that when I left the program running through the night, only a few people downloaded anything - yet if I connect with my static IP without VPN and with ports actually open, I get 5x that number of people starting to download stuff in 10-15minute window alone. 

Any feedback/suggestions would be welcome - I think I tried the usual troubleshooting tips already though - did try binding the interface to Proton VPN and unbinding, checking upnp and unchecking etc (gives out _**UPnP: Failed to forward external port XXXXX: No UPnP devices found**_ error if checked, behaviour itself doesnt seem to change at all though)

**Comment by slook** [comment]
You have outlined a reason why changes to the NAT-PMP and UPnP modules are required, because I don't think the NAT-PMP/UPnP function is capable of dealing with interface binding, it only gets forwarding route from a default gateway, so this could lead to a conflict, in such a case of there being a virtual private network interface, whereby the desired route is not on the system default gateway.

I was working on this the other day to make it respect the binding interface setting, see https://github.com/slook/nicotine-plus/commit/435f30f9aa065a26ec34133dd09fcdbd9e1d1532 in my development branch:
https://github.com/slook/nicotine-plus/tree/portmapper-gateway-fallback-netstat

However I am unable to fully test it, because I am behind a public NAT for which I do not have any possibility of port forwarding. Are you able to check out my development branch? You could run it from a local Git folder, or you can install an Artifact from packaging run at link:
https://github.com/slook/nicotine-plus/actions/runs/8214087851

For these testing and troubleshooting tasks it will be useful to activate the "Miscellaneous" log category in [debug logging](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/DEVELOPING.md#debug-logging), it will reveal more details while establishing an initial connection with your router, or your providers virtual router.

Alternatively, you might have to turn the NAT-PMP/UPnP option off in Preferences > Network, in order to avoid it from interfering with your manual port forwarding router settings.

**Comment by exostalker** [comment, marked as answer]
I tested that ^ and sadly same thing. Also same for turning off NAT-PMP/UPnP, already tried it, same with it enabled or disabled.

**Reply by slook** [reply]
Did you test it with a network interface binding in Nicotine+ Preferences? That is the only case where it might make any difference.

Also with the extra debug logging in my development build, perhaps you can verify exactly which interface is used.

**Reply by slook** [reply]
Please open a terminal and provide the outputs of the following commands:

- `cat /proc/net/route`
- `netstat -rn`

**Reply by slook** [reply]
`netstat -rn`

**Reply by exostalker** [reply]
Yeah that above is the "netstat -rn"

**Reply by slook** [reply]
The code in my development version is flawed, it won't work on Windows. A different approach will be required.

For now it is probably necessary to disable UPnP/NAT-PMP and specify an interface binding adapter in Preferences.

Additionally, some other configuration may be required in the OS, such as firewall or routing of the forwarded port.

**Comment by toineenzo** [comment]
If you're running N+ inside Docker, make sure to check out https://github.com/qdm12/gluetun . It makes it really easy to correctly implement VPN with docker container and N+
